### PR TITLE
feature: node cache/storage expiration

### PR DESCRIPTION
**Pull request by comountainclimber**
**What current issue(s) from Trello/Github does this address?**
This PR begins the iterative process of issue/enhancement https://github.com/CityOfZion/neon-wallet/issues/1712 - Specifically it adds an expiration to the "best node" algo and also sets an expiration on manually set nodes... This will resolve a number of _possible_ latent issues around the current set up

**What problem does this PR solve?**
Because nodes could potentially fall behind go offline etc leaving them in memory (or local storage) permanently is generally a bad idea

**How did you solve this problem?**
node url is set in storage with a timestamp which is always compared to the current time... if its expired the node selection algorithm is run

**How did you make sure your solution works?**
By manual and automated test coverage

**Are there any special changes in the code that we should be aware of?**

**Is there anything else we should know?**

- [x] Unit tests written? 😎

### CODE REVIEW

File: app/actions/nodeStorageActions.js [hunk 1]
```diff
@@ -81,14 +108,20 @@ export const getRPCEndpoint = async (
...
+export const getNode = async (net: Net): Promise<string> => {
   const storage = await getStorage(`${STORAGE_KEY}-${net}`).catch(console.error)
-  if (!storage) return ''
-  return isEmpty(storage) ? '' : storage
+  const nodeInStorage = get(storage, 'node')
+  const expiration = get(storage, 'timestamp')
+  if (!nodeInStorage) return ''
+  if (!expiration) return ''
+  if (determineIfCacheIsExpired(expiration)) {
+    return ''
```
Comment (evgenyboxer) [inline app/actions/nodeStorageActions.js]: these expressions can be combined into one `if (! nodeInStorage || !expiration || !determineIfCacheIsExpired(expiration)) { return '' }`

File: flow-typed/declarations.js [hunk 2]
```diff
@@ -28,6 +29,8 @@ declare type ReduxAction = () => {
 
 declare type NetworkType = string
 
+declare type NetworkLabelTypes = $Values<typeof NETWORK_LABELS>
```
Comment (evgenyboxer) [inline flow-typed/declarations.js]: hmm is this actually working? thinking if this should just be `MAIN_NETWORK_LABEL | TEST_NETWORK_LABEL`

File: app/actions/nodeStorageActions.js [hunk 3]
```diff
@@ -15,19 +15,44 @@ import { findNetworkByLabel } from '../core/networks'
...
+
+export const parseUrlData = (data: {
+  height: number,
+  port: string,
+  protocol: string,
+  url: string
+}): string => {
+  let url = data.protocol ? `${data.protocol}://${data.url}` : data.url
+  url = data.port ? `${url}:${data.port}` : url
+  return url
```
Comment (evgenyboxer) [inline app/actions/nodeStorageActions.js]: I think the function name is a bit confusing, it's more like `buildURL`

Regarding the implementation, how about using lodash compact and an array that you join? (https://lodash.com/docs/4.17.11#compact)


```
const { protocol, url, port } = url

return compact([
  protocol && `${protocol}://`,
  url,
  port && `:${port}`
]).join('')
```

File: flow-typed/declarations.js (hunk 2 above)
Comment (comountainclimber) [reply flow-typed/declarations.js]: When using this syntax initially there was previously an error that seems to have disappeared ✨ thanks @evgenyboxer

**Comment by comountainclimber** [comment]
Thanks for the solid review @evgenyboxer 💪

**Comment by comountainclimber** [comment]
Thanks @dvdschwrtz 😄

**Pull request merged**